src/main/kotlin/io/perfometer/dsl/HttpDsl.kt

```diff
@@ -1,117 +1,130 @@
+import io.perfometer.runner.CoroutinesScenarioRunner
+import io.perfometer.runner.ScenarioRunner
+import io.perfometer.statistics.ScenarioStatistics
 import java.net.URL
 import java.time.Duration
 import java.util.*
 
+sealed class HttpStep
+data class RequestStep(val request: HttpRequest) : HttpStep()
+data class PauseStep(val duration: Duration) : HttpStep()
+
 typealias HttpHeader = Pair<String, String>
 typealias HttpParam = Pair<String, String>
 
-class RequestBuilder(
-        val url: URL,
-        val method: HttpMethod,
-        initialHeaders: List<() -> HttpHeader>
+class RequestDsl(
+        private val url: URL,
+        private val method: HttpMethod,
+        initialHeaders: List<HttpHeader>
 ) {
-    private var path: () -> String = { "" }
-    private val headers: MutableList<() -> HttpHeader> = initialHeaders.toMutableList()
-    private val params: MutableList<() -> HttpParam> = mutableListOf()
-
-    var body: () -> ByteArray = { ByteArray(0) }
-        private set
+    private var path: String = ""
+    private val headers: MutableList<HttpHeader> = initialHeaders.toMutableList()
+    private val params: MutableList<HttpParam> = mutableListOf()
+    private var body: ByteArray = ByteArray(0)
+    private var consumer: (HttpResponse) -> Unit = {}
 
-    var consumer: (HttpResponse) -> Unit = {}
-        private set
-
-    fun path(path: () -> String): RequestBuilder {
+    fun path(path: String) {
         this.path = path
-        return this
     }
 
-    fun body(body: () -> ByteArray): RequestBuilder {
+    fun body(body: ByteArray) {
         this.body = body
-        return this
     }
 
-    fun header(header: () -> HttpHeader): RequestBuilder {
+    fun header(header: HttpHeader) {
         headers.add(header)
-        return this
     }
 
-    fun param(param: () -> HttpParam): RequestBuilder {
+    fun param(param: HttpParam) {
         params.add(param)
-        return this
     }
 
-    fun consume(consumer: (HttpResponse) -> Unit): RequestBuilder {
+    fun consume(consumer: (HttpResponse) -> Unit) {
         this.consumer = consumer
-        return this
     }
 
-    fun pathWithParams(): String {
-        return path() + paramsToString()
+    private fun pathWithParams(): String {
+        return path + paramsToString()
     }
 
     private fun paramsToString(): String {
         return if (this.params.isNotEmpty())
-            this.params.map { it() }.joinToString("&", "?") { "${it.first}=${it.second}" }
+            this.params.joinToString("&", "?") { "${it.first}=${it.second}" }
         else ""
     }
 
-    fun headers(): Map<String, String> {
-        return this.headers.map { it() }
+    private fun headers(): Map<String, String> {
+        return this.headers
                 .groupBy({ it.first }, { it.second })
                 .mapValues { it.value.joinToString(",") }
     }
+
+    fun build(): HttpRequest = HttpRequest(method, url, pathWithParams(), headers(), body, consumer)
```

Comment:
A minor thing: if the type is obvious, I wouldn't use type declaration in the signature, but simply go with: `fun build() = HttpRequest(...)`

Reply:
done